### PR TITLE
[cudagraph] Capture single-stage pipeline forward-backward steps

### PR DESCRIPTION
**Pull request by xmfan**
Enable CUDA graph capture and replay for pipeline schedules with one stage per rank, including GPipe and 1F1B. Capture the forward-backward schedule while keeping input preprocessing and token accounting outside the captured region.

This is a manually squashed consolidation of the previous stack and its review follow-ups:

- [#4432: Structured CUDA graph inputs](https://github.com/pytorch/torchtitan/pull/4432)
- [#4433: Warm up two optimizer steps before capture](https://github.com/pytorch/torchtitan/pull/4433)
- [#4435: Single-stage pipeline CUDA graph capture](https://github.com/pytorch/torchtitan/pull/4435)

The original PRs were submitted using stack-pr. After losing upstream write access, updating and landing that stack became difficult; this PR replaces those three unmerged PRs with one commit targeting `main` from `xmfan:cudagraph-review-followups`. It builds on the already merged [#4430: Release consumed pipeline loss graphs](https://github.com/pytorch/torchtitan/pull/4430).

The wrapper handles structured positional and keyword inputs, checks stable input structure and tensor metadata, and copies new tensor values into capture-owned buffers. Warmup invocation counting lives in `wrap_with_cuda_graph` and accounts for gradient accumulation and finite or continuous SDC replay schedules. As requested in review, `gradient_accumulation_steps`, `sdc_num_steps`, and `sdc_num_replays` are required keyword arguments. Two optimizer steps remain a conservative warmup default, allowing an eager step after lazy optimizer state initialization.

Pipeline capture reuses a device-resident loss sentinel on non-last stages. Configuration rejects looped schedules and pipeline validation, which reinitializes the shared schedule. The existing real-process-group 1F1B and PP+FSDP integration recipes now enable CUDA graphs and run 10 training steps.

Validation: CI

TorchFT changes and the separate CPU override-registry fix remain outside this PR's scope.

### CODE REVIEW

**Comment by xmfan** [comment]
@tianyu-l I don't have permissions to merge the stack 😂

File: torchtitan/distributed/cudagraph.py [hunk 1]
```diff
@@ -354,18 +354,31 @@ def teardown(self) -> None:
...
 
 
+# TODO: Unify PP and non-PP callable signatures to restore strict input typing.
 def wrap_with_cuda_graph(
-    fn: Callable[..., torch.Tensor], *, num_warmup_iterations: int = 1
+    fn: Callable[..., torch.Tensor],
+    *,
+    gradient_accumulation_steps: int = 1,
+    sdc_num_steps: int = 0,
+    sdc_num_replays: int = 0,
```
Comment (tianyu-l) [inline torchtitan/distributed/cudagraph.py]: these shouldn't have default

**Review by tianyu-l — APPROVED** [review]
I couldn't merge previous stack either because they used ghstack. Maybe could you submit a consolidated one to ease the merge.

**Comment by xmfan** [comment]
Put everything in this PR
